<a href="https://colab.research.google.com/github/jorgevillegazz/Challange-Telecom-X/blob/main/TelecomX_LATAM_de_George.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#💡Acerca del desafío 💡

Descripción

# ***Telecom X - Análisis de Evasión de Clientes***

Has sido contratado como asistente de análisis de datos en Telecom X y formarás parte del proyecto "Churn de Clientes". **La empresa** enfrenta una alta tasa de cancelaciones y **necesita comprender los factores que llevan a la pérdida de clientes.**

Tu desafío será **recopilar, procesar y analizar los dato**s, utilizando Python y sus principales bibliotecas **para extraer información valiosa**. A partir de tu análisis, el equipo de Data Science podrá avanzar en modelos predictivos y desarrollar estrategias para reducir la evasión.

#### Diccionario de datos

- `customerID`: número de identificación único de cada cliente
- `Churn`: si el cliente dejó o no la empresa
- `gender`: género (masculino y femenino)
- `SeniorCitizen`: información sobre si un cliente tiene o no una edad igual o mayor a 65 años
- `Partner`: si el cliente tiene o no una pareja
- `Dependents`: si el cliente tiene o no dependientes
- `tenure`: meses de contrato del cliente
- `PhoneService`: suscripción al servicio telefónico
- `MultipleLines`: suscripción a más de una línea telefónica
- `InternetService`: suscripción a un proveedor de internet
- `OnlineSecurity`: suscripción adicional de seguridad en línea
- `OnlineBackup`: suscripción adicional de respaldo en línea
- `DeviceProtection`: suscripción adicional de protección del dispositivo
- `TechSupport`: suscripción adicional de soporte técnico, menor tiempo de espera
- `StreamingTV`: suscripción de televisión por cable
- `StreamingMovies`: suscripción de streaming de películas
- `Contract`: tipo de contrato
- `PaperlessBilling`: si el cliente prefiere recibir la factura en línea
- `PaymentMethod`: forma de pago
- `Charges.Monthly`: total de todos los servicios del cliente por mes
- `Charges.Total`: total gastado por el cliente

#📌 Extracción

##Cargando Librerias

In [ ]:
# Importamos las librerías
import pandas as pd
import requests
import matplotlib.pyplot as plt
import seaborn as sns

##Importando Datos

In [ ]:
url = "https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/refs/heads/main/TelecomX_Data.json"
df_telecom = pd.read_json(url)

# Visualizando los Datos

In [ ]:
df_telecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [ ]:
df_telecom

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

# El archivo **json**  que proporciona la **Empresa** tiene cuatro columnas directorio la cuales  se tiene  que proceceder a **normalizar** (customer, phone, internet y account)

***


**Primero SEPARAMOS las columnas ID y Churn**  de las demas columnas que tienen diccionario

In [ ]:
# Extraer columnas customerID	Churn
df_customerID_Churn = df_telecom[['customerID', 'Churn']]

Extraemos las 4 columnas diccionario  para proceder a **normalizar**

In [ ]:
# Extraer columnas que  tienen directorio y normalizarlos
df_customer = pd.json_normalize(df_telecom['customer'])
df_phone = pd.json_normalize(df_telecom['phone'])
df_internet = pd.json_normalize(df_telecom['internet'])
df_account = pd.json_normalize(df_telecom['account'])

Juantamos todas las columnas  para crea un **nuevo df**

---



In [ ]:
# Se concatena o juentan las columnas para crear el nuevo df  ya normalizado o sin direcotrios
df_normalizado = pd.concat([df_customerID_Churn, df_customer, df_phone, df_internet, df_account], axis=1)

#Obligamos a mostrar  TODAS la columnas con la siguinete lienea de  comando de pandas 'display.max_columns', None
pd.set_option('display.max_columns', None)

Es hora de ver como quedo nuestra  Nueva Data Frame  ya formalizada




In [ ]:
# Obligamos a mostrar TODAS la columnas con la siguinete linea de comando de
# pandas 'display.max_columns', None
pd.set_option('display.max_columns', None)
df_normalizado

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService OnlineSecurity  \
0             Yes                No             DSL             No   
1             Yes               Yes             DSL             No   
2             Yes                No     Fiber optic             No   
3             Yes                No     Fiber optic             No   
4             Yes                No     Fiber optic             No   
...           ...               ...             ...            ...   
7262          Yes                No             DSL            Yes   
7263          Yes               Yes     Fiber optic             No   
7264          Yes                No             DSL             No   
7265          Yes                No             DSL            Yes   
7266           No  No phone service             DSL            Yes   

     OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
0             Yes               No         Yes         Yes              No   
1              No               No          No          No             Yes   
2              No              Yes          No          No              No   
3             Yes              Yes          No         Yes             Yes   
4              No               No         Yes         Yes              No   
...           ...              ...         ...         ...             ...   
7262           No               No         Yes          No              No   
7263           No               No          No          No             Yes   
7264          Yes               No          No          No              No   
7265           No              Yes         Yes          No             Yes   
7266          Yes              Yes          No         Yes             Yes   

            Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0           One year              Yes      Mailed check            65.60   
1     Month-to-month               No      Mailed check            59.90   
2     Month-to-month              Yes  Electronic check            73.90   
3     Month-to-month              Yes  Electronic check            98.00   
4     Month-to-month              Yes      Mailed check            83.90   
...              ...              ...               ...              ...   
7262        One year               No      Mailed check            55.15   
7263  Month-to-month              Yes  Electronic check            85.10   
7264  Month-to-month              Yes      Mailed check            50.30   
7265        Two year               No      Mailed check            67.85   
7266        Two year               No  Electronic check            59.00   

     Charges.Total  
0            593.3  
1            542.4  
2           280.85  
3          1237.85  
4            267.4  
...            ...  
7262         742.9  
7263        1873.7  
7264         92.75  
7265       4627.65  
7266        3707.6  

[7267 rows x 21 columns]

In [ ]:
df_normalizado.sample(10)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
2988  4140-MUHUG   Yes  Female              1      No         No       3   
1601  2272-UOINI    No  Female              0      No         No       7   
3239  4523-WXCEF   Yes  Female              0     Yes         No      59   
4771  6537-QLGEX    No  Female              0      No         No      16   
4600  6303-KFWSL    No  Female              0      No         No      29   
3723  5134-IKDAY   Yes  Female              0     Yes        Yes       1   
912   1289-RKJJR          Male              0      No         No      28   
5362  7321-PKUYW   Yes  Female              0      No         No      11   
2940  4083-BFNYK    No  Female              1     Yes         No      38   
3503  4828-FAZPK    No  Female              0     Yes        Yes      16   

     PhoneService     MultipleLines InternetService OnlineSecurity  \
2988          Yes                No     Fiber optic             No   
1601          Yes                No             DSL            Yes   
3239          Yes               Yes     Fiber optic             No   
4771           No  No phone service             DSL            Yes   
4600          Yes               Yes             DSL             No   
3723          Yes                No     Fiber optic             No   
912           Yes                No     Fiber optic             No   
5362          Yes               Yes     Fiber optic             No   
2940          Yes               Yes     Fiber optic            Yes   
3503          Yes               Yes     Fiber optic             No   

     OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
2988           No              Yes          No         Yes              No   
1601          Yes               No         Yes         Yes             Yes   
3239          Yes               No          No         Yes             Yes   
4771           No               No          No         Yes              No   
4600          Yes               No          No          No              No   
3723           No               No          No          No              No   
912            No              Yes         Yes          No             Yes   
5362           No               No         Yes         Yes              No   
2940          Yes               No          No          No             Yes   
3503          Yes               No          No          No              No   

            Contract PaperlessBilling            PaymentMethod  \
2988  Month-to-month              Yes         Electronic check   
1601  Month-to-month              Yes         Electronic check   
3239  Month-to-month               No         Electronic check   
4771  Month-to-month              Yes         Electronic check   
4600        One year              Yes         Electronic check   
3723  Month-to-month              Yes         Electronic check   
912   Month-to-month               No         Electronic check   
5362  Month-to-month               No         Electronic check   
2940        One year               No  Credit card (automatic)   
3503  Month-to-month              Yes  Credit card (automatic)   

      Charges.Monthly Charges.Total  
2988            86.85        220.95  
1601            78.50        571.05  
3239           101.10        6039.9  
4771            38.90         664.4  
4600            55.25        1620.2  
3723            69.80          69.8  
912             91.00       2626.15  
5362            90.15        987.95  
2940            95.00       3591.25  
3503            79.95       1267.95

#🔧 Transformación

✅ Es Hora Explorar las columnas del dataset y verificar sus tipos de datos.

In [ ]:
#visualizamos el tipo de dato por columna
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


Se Observa que nuestro Data Frame  contiene **7267 registros** con un total de **21 Columnas.**

**La mayoria son de tipo Object**,  con forme avanzemos veremos que variables cambiaremos su tipo.

Tambien se observa que hay 2 columnas que tienen un **punto "."**  en su nombre, los cuales serán remplazados por **"_" (guión bajo)**
***

In [ ]:
# Limpiar nombres de columnas que tienen punto (Charges.Monthly y Charges.Total)
df_normalizado.columns = [col.replace('Charges.', 'Charges_')  for col in df_normalizado.columns]

# visualizar resultado
df_normalizado.head()


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService OnlineSecurity OnlineBackup  \
0          Yes            No             DSL             No          Yes   
1          Yes           Yes             DSL             No           No   
2          Yes            No     Fiber optic             No           No   
3          Yes            No     Fiber optic             No          Yes   
4          Yes            No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod  Charges_Monthly Charges_Total  
0              Yes      Mailed check             65.6         593.3  
1               No      Mailed check             59.9         542.4  
2              Yes  Electronic check             73.9        280.85  
3              Yes  Electronic check             98.0       1237.85  
4              Yes      Mailed check             83.9         267.4

In [53]:
#Verificacion del cambio de nombre en las columnas que tenian punto
columnas=list(df_normalizado.columns)
columnas
#df_normalizado.columns.tolist()

['customerID',
 'Churn',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Charges_Monthly',
 'Charges_Total']

#**Conversion de Datos a Tipo Numérico**

De momento creo  que una de las columnas importantes **Charges_Total** la convertiremos de tipo objet a tipo float para nuestrofuturo  analisis exploratorio.

Mas adelante  veremos cuantos mas  cambiaremos

In [40]:
from numpy import astype


# df_telecom_norm['Charges.Total']=df_telecom_norm['Charges.Total'].astype('float64')
# la anterior linea me causaba error. tube que usar el siguiente codico con la funcion  pd.to_numeric con parametro, error='çoerce'
df_normalizado['Charges_Total'] =pd.to_numeric(df_normalizado['Charges_Total'], errors='coerce')



In [41]:
# Verificamos el nuevo tipos de datos
df_normalizado.dtypes[['Charges_Total']]


Charges_Total    float64
dtype: object

#**Comprobación de incoherencias en los datos**
Busqueda de valores Nulos (Nan)

In [65]:
# CREAMOS un ciclo for para detectar valores  faltantes (NaN)
for nombre in columnas:
    nulos=(df_normalizado[nombre].isna().sum())
    if nulos > 0:
      print(f"\033[1m{nombre}\033[0m",'.....',nulos)

Charges_Total ..... 11


Observamos  que Chargues_Total tiene 11 registros NaN

In [66]:
df_normalizado[df_normalizado['Charges_Total'].isna()]


customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
975   1371-DWPAZ    No  Female              0     Yes        Yes       0   
1775  2520-SGTTA    No  Female              0     Yes        Yes       0   
1955  2775-SEFEE    No    Male              0      No        Yes       0   
2075  2923-ARZLG    No    Male              0     Yes        Yes       0   
2232  3115-CZMZD    No    Male              0      No        Yes       0   
2308  3213-VVOLG    No    Male              0     Yes        Yes       0   
2930  4075-WKNIU    No  Female              0     Yes        Yes       0   
3134  4367-NUYAO    No    Male              0     Yes        Yes       0   
3203  4472-LVYGI    No  Female              0     Yes        Yes       0   
4169  5709-LVOEQ    No  Female              0     Yes        Yes       0   
5599  7644-OMVMY    No    Male              0     Yes        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
975            No  No phone service             DSL                  Yes   
1775          Yes                No              No  No internet service   
1955          Yes               Yes             DSL                  Yes   
2075          Yes                No              No  No internet service   
2232          Yes                No              No  No internet service   
2308          Yes               Yes              No  No internet service   
2930          Yes               Yes             DSL                   No   
3134          Yes               Yes              No  No internet service   
3203           No  No phone service             DSL                  Yes   
4169          Yes                No             DSL                  Yes   
5599          Yes                No              No  No internet service   

             OnlineBackup     DeviceProtection          TechSupport  \
975                   Yes                  Yes                  Yes   
1775  No internet service  No internet service  No internet service   
1955                  Yes                   No                  Yes   
2075  No internet service  No internet service  No internet service   
2232  No internet service  No internet service  No internet service   
2308  No internet service  No internet service  No internet service   
2930                  Yes                  Yes                  Yes   
3134  No internet service  No internet service  No internet service   
3203                   No                  Yes                  Yes   
4169                  Yes                  Yes                   No   
5599  No internet service  No internet service  No internet service   

              StreamingTV      StreamingMovies  Contract PaperlessBilling  \
975                   Yes                   No  Two year               No   
1775  No internet service  No internet service  Two year               No   
1955                   No                   No  Two year              Yes   
2075  No internet service  No internet service  One year              Yes   
2232  No internet service  No internet service  Two year               No   
2308  No internet service  No internet service  Two year               No   
2930                  Yes                   No  Two year               No   
3134  No internet service  No internet service  Two year               No   
3203                  Yes                   No  Two year              Yes   
4169                  Yes                  Yes  Two year               No   
5599  No internet service  No internet service  Two year               No   

                  PaymentMethod  Charges_Monthly  Charges_Total  
975     Credit card (automatic)            56.05            NaN  
1775               Mailed check            20.00            NaN  
1955  Bank transfer (automatic)            61.90            NaN  
2075               Mailed check            19.70            NaN  
2232               Mailed check            20.25            NaN  
2308  

Ahora  buscaremos  datos  que tengan valor datos vacios' '


In [64]:
# CREAMOS un ciclo for para detectar datos vacios en cada  columna
for nombre in df_normalizado.columns:
    nulos=(df_normalizado[nombre]=='').sum()
    if nulos > 0:
      print(f"\033[1m{nombre}\033[0m",'.....',nulos)

Churn ..... 224


Interesante saber que Churn cuenta  con 224 registros vacios los cuales en breve los eliminaremos de nuestro Data Frame
***

Verificamos cuantos Clientes nuevos hay ( cero meses)

In [67]:
df_normalizado.tenure.value_counts()

tenure
1     634
72    369
2     246
3     207
4     185
     ... 
28     60
39     59
44     54
36     50
0      11
Name: count, Length: 73, dtype: int64

In [69]:
# REVISAMOS cuantos registros de clientes han cancelado y cuantos siguen activos
df_normalizado.Churn.value_counts()

Churn
No     5174
Yes    1869
        224
Name: count, dtype: int64

#**Manejo de inconsistencias**
Reemplazo de Valores nulos (Nan)

In [70]:
# Reemplazar valores nulos en Charges_Total con 0
# Los clientes con tenure(meses de contrato) = 0 consideré que aún no han generado cobros

df_normalizado['Charges_Total'] = df_normalizado['Charges_Total'].fillna(0)

Revisamos el cambio

In [71]:
df_normalizado.Charges_Total.isna().sum()

np.int64(0)

In [ ]:
#Comprobando los cambios
df_normalizado[['Charges_Monthly', 'Charges_Total']].isna().sum()


Charges_Monthly    0
Charges_Total      0
dtype: int64

In [ ]:
columnas=list(df_normalizado.columns)
columnas

['customerID',
 'Churn',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Charges.Monthly',
 'Charges.Total']

#Buscamos si hay **Strings vacios** ,**Valores Nulos** y  con **Espacios en Blanco** en cada  columna


In [ ]:
# CREAMOS un ciclo for para detectar datos nulos en cada  columna
print("--- Iniciando impresión de columnas ---")
for nombre in df_normalizado.columns:
    nulos=(df_normalizado[nombre]=='').sum()
    if nulos > 0:
      print(nombre,'......',nulos)
print("--- Fin de la impresión ---")

--- Iniciando impresión de columnas ---
Churn ...... 224
--- Fin de la impresión ---


In [ ]:
# 2. CREAMOS UN ciclo for para DETECTAR espacios en blanco en cada columna
print("--- Iniciando impresión de columnas con espacios---")
for nombre in df_normalizado.columns:
    espacios=(df_normalizado[nombre]==' ').sum()
    if espacios > 0:
        print(nombre,'......',espacios)
print("--- Fin de la impresión ---")

--- Iniciando impresión de columnas con espacios---
Charges.Total ...... 11
--- Fin de la impresión ---


--- Iniciando impresión de columnas con espacios---
--- Fin de la impresión ---


In [ ]:
#ahora trataremos los valores str vacios, considere que los valores en churn no los vamos a eliminar por el momento, a la columna churn la trataremos como valores enteros
#pasamos el churn a valores si 1, no 0 y los vacios -1 para futuros analisis
df_normalizado['Churn']= df_normalizado['Churn'].str.strip().replace('',-1)
#para el changes.total que seria lo que gasto el cliente en total, lo dejaremos en 0 porque puede que el cliente sea nuevo o en cierto periodo no facturo
df_normalizado['Charges.Total']= df_normalizado['Charges.Total'].str.strip().replace(' ','0')
df_normalizado.isnull().sum()

customerID            0
Churn               224
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod         0
Charges.Monthly       0
Charges.Total         0
dtype: int64

In [48]:
columnas=list(df_normalizado.columns)
columnas

['customerID',
 'Churn',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Charges_Monthly',
 'Charges_Total']

In [50]:
# CREAMOS un ciclo for para detectar datos nulos en cada  columna
print("--- Iniciando impresión de columnas ---")
for nombre in columnas:
    nulos=(df_normalizado[nombre].isna().sum())
    if nulos > 0:
      print(nombre,'......',nulos)
print("--- Fin de la impresión ---")


--- Iniciando impresión de columnas ---
Charges_Total ...... 11
--- Fin de la impresión ---


In [52]:
# 2. CREAMOS UN ciclo for para DETECTAR espacios en blanco en cada columna
print("--- Iniciando impresión de columnas con espacios---")
for nombre in columnas:
    espacios=(df_normalizado[nombre]=='').sum()
    if espacios > 0:
        print(nombre,'......',espacios)
print("--- Fin de la impresión ---")

--- Iniciando impresión de columnas con espacios---
Churn ...... 224
--- Fin de la impresión ---


In [ ]:
#ahora trataremos los valores str vacios, considere que los valores en churn no los vamos a eliminar por el momento, a la columna churn la trataremos como valores enteros
#pasamos el churn a valores si 1, no 0 y los vacios -1 para futuros analisis
df_telecom_norm['Churn']= df_telecom_norm['Churn'].str.strip().replace('',-1)
#para el changes.total que seria lo que gasto el cliente en total, lo dejaremos en 0 porque puede que el cliente sea nuevo o en cierto periodo no facturo
df_telecom_norm['Charges.Total']= df_telecom_norm['Charges.Total'].str.strip().replace('','0')
df_telecom_norm.isnull().sum()

customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

In [ ]:
df_normalizado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


In [ ]:
#Verificamos duplicados
df_normalizado.duplicated().sum()

np.int64(0)

#📊 Carga y análisis

#📄Informe final